keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1067 @@
+  (e.g. kubectl apply -f <lots of yaml files>) only works if the apply
+  operations are ordered and objects are verified to exist at specific steps.
+
+#### Additional Capabilities (WIP)
+
+In addition to the high level API structure proposed above, there are quite
+a few capatiblities that are important to this enhancement that we are working
+on defining. These are all work-in-progress.
+
+Rule Message:
+
+- message formatting (need support for string formatting. Allow a CEL expression
+  to be used to format the whole string? Or allow CEL expressions to evaluate
+  template values?)
+- severity (e.g. "Warning")
+- status type equivalent (e.g. Forbidden can be returned from a webhook using
+  HTTP 403) (can a match rule override this? or can it be selected via a CEL
+  expression?)
+- Similar to how each match has configuration, can each match also select
+  the severity of messages?
+
+Rule scope:
+
+- Dereferencing to deeply nested fields in CEL can be burdonsome. A field
+  containing a field path would allow a CEL rule to be scope to some node in
+  the schema.
+- Also, the idea of named scopes (e.g. "ALL_CONTAINERS") would allow a single
+  rule to be applied to all container definitions in a resource (e.g. both
+  "containers" and "initContainers" in a pod)
+
+Failure Policy:
+
+- Not the same as webhooks, where failure policy is about the remote request
+- What if a CEL rule evaluates to an error?
+- What if a CEL rule is for a CRD kind and the CRD changes such that the CEL
+  rule fails type checking? (or the CRD does not exist?)
+
+Decision Rules:
+
+- Are overlapping matches combined using an OR or AND? Can this be configured?
+- Can a validation rule depend on another validation rule (only evaluate if the
+  other rule is true?)
+- How to make it so that an "not authorized" error prevents all other validation
+  messages from being returned (to prevent exfiltration)?
+
+Composition utilities:
+
+- Ability to define a sub-expression and then use it in validation rules (xref
+  cel-policy-template "terms")
+
+Transition rules:
+
+- Should we use "self" and "oldSelf" ? How do transition rules work for delete?
+
+Access to namespace labels:
+
+- Most heavily needed fields not directly available in the resource being
+  validated. Note that namespaceSelectors already allow matches to examine
+  namespace levels.
+
+Type safety:
+
+- What happens when a ValidatingAdmissionConfiguration is created in a cluster
+  where either CRDs or native types are incompatible with CEL rules and the CEL
+  rules fail to compile?
+  - Option: Put something into status about the type check error and apply the
+    fail policy instead of evaluating the CEL expression at runtime
+  - Option: Put something into status and run the CEL expression dynamically, if
+    it fails with an error apply the fail policy
+  - Option: Provide a field on each rule that determines the behavior
+    ("TypeCheckFailPolicy: (FailAll|RunDynamic)").
+
```

Comment:
a section on interaction with aggregated API servers would be useful

1. the API server validating/persisting the ValidatingAdmissionConfiguration instances isn't the same one serving aggregated types, so wouldn't necessarily have schema info to check type safety
2. the aggregated API server is responsible for enforcing admission on its custom types, so the implementation that reads ValidatingAdmissionConfiguration instances and enforces them would have to live in k8s.io/apiserver and be active in aggregated API servers to enforce admission on aggregated types effectively (same as admission webhooks today)

Reply:
Thanks! I've added a section. It's still marked as TODO but it will make sure we track this.  Throws a wrench into type safety for sure, I need to think about that one.